### PR TITLE
Add new route for import result csv

### PR DESCRIPTION
**Pull request by chukarave**
This pr adds the route and introduces the
`ImportsResultsController`, which uses the new `ImportResults` service to generate the CSV dynamically. 
This also adds a new config section with the file name template and the headers.

Bug: T325158

### CODE REVIEW

File: tests/Feature/ImportReportTest.php [hunk 1]
```diff
@@ -21,9 +21,7 @@ class ImportReportTest extends TestCase
     public function test_generate_csv_writes_to_path(): void
     {
         $user = User::factory()->uploader()->create();
-        $import = ImportMeta::factory()->for($user)->create(
-            ['status' => 'completed']
-        );
+        $import = ImportMeta::factory()->for($user)->create();
```
Comment (lucaswerkmeister) [inline tests/Feature/ImportReportTest.php:24]: Is this change needed? The tests still seem to pass without it.

File: tests/Feature/ImportResultsTest.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            ->for($import)
+            ->create();
+
+        Storage::fake('local');
+        $filename = 'temp_test.csv';
+
+        Storage::put($filename, '');
+        $path = Storage::path($filename);
+
+        $this->travelTo(now()); // stop the clock
```
Comment (lucaswerkmeister) [inline tests/Feature/ImportResultsTest.php:35]: Do we need the time travel? As far as I can tell, nothing in the CSV depends on it, and the test still seems to pass without it.

File: app/Http/Controllers/ImportsResultsController.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+namespace App\Http\Controllers;
+
+use App\Services\ImportResults;
+
+class ImportsResultsController extends Controller
+{
+    public function showResultsCsv(int $import_id)
+    {
+        $filename = strtr(config('imports.results.filename_template'), [
+            ':datetime' => now()->format('Y-m-d_H-i-s')
```
Comment (lucaswerkmeister) [inline app/Http/Controllers/ImportsResultsController.php:12]: I’d use colons instead of dashes in the “time” part, and preferably also a `T` instead of `_` to separate the “date” and “time” parts, to match ISO 8601.

File: app/Services/ImportResults.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                $mismatch['wikidata_value'],
+                $mismatch['meta_wikidata_value'],
+                $mismatch['external_value'],
+                $mismatch['external_url'],
+                $mismatch['review_status']
+            ];
+        }, $mismatches->toArray());
+
+        // Add column keys as CSV headers
+        $mismatchesWithKeys = array_merge([config('imports.results.column_keys')
```
Comment (lucaswerkmeister) [inline app/Services/ImportResults.php:34]: IMHO it would be nicer to emit the headers in a separate `fputcsv()` call (like in `ImportReport`), rather than merge two fairly different arrays together.

File: app/Services/ImportResults.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        $mismatches = Mismatch::get();
+
+        // Creating the output stream
+        $handle = fopen($path, 'w');
+
+        $import = ImportMeta::find($import_id);
+        $mismatches = Mismatch::whereBelongsTo($import)->get();
+
+        $mismatchValues = array_map(function ($mismatch) {
```
Comment (lucaswerkmeister) [inline app/Services/ImportResults.php:20]: I think instead of  mapping the whole result at once, and then iterating over the resulting array, we could also iterate over the mismatches directly and call `fputcsv()` for each row?

**Comment by lucaswerkmeister** [comment]
By the way, this pull request also already implements T325177.

File: tests/Feature/ImportReportTest.php (hunk 1 above)
Comment (chukarave) [reply tests/Feature/ImportReportTest.php:24]: It was left over from the test I took the code from but was redundant for this test therefore removed

Comment (chukarave) [reply tests/Feature/ImportReportTest.php:24]: yeah this seems to be unintentional

File: tests/Feature/ImportResultsTest.php (hunk 2 above)
Comment (chukarave) [reply tests/Feature/ImportResultsTest.php]: removed

File: app/Services/ImportResults.php (hunk 5 above)
Comment (chukarave) [reply app/Services/ImportResults.php]: I ended up making this a LazyCollection, where the mapping is only triggered by the `.each()` iteration which writes the mismatch to the csv

File: tests/Feature/ImportReportTest.php [hunk 6]
```diff
@@ -46,8 +46,6 @@ public function test_generate_csv_writes_to_path(): void
         $total = $mismatches->count();
         $pending = $mismatches->where('review_status', 'pending')->count();
 
-        $this->travelTo(now()); // stop the clock
```
Comment (itamargiv) [inline tests/Feature/ImportReportTest.php:49]: This change doesn't necessarily relate to the ticket we are working, I think it will probably be best to do this in a separate PR.

File: config/imports.php [hunk 7]
```diff
@@ -51,5 +51,19 @@
             "expiry date",
             "expired?"
         ]
+    ],
+
+    'results' => [
+        'filename_template' => 'import-results-:datetime.csv',
```
Comment (itamargiv) [inline config/imports.php:57]:
```suggestion
        'filename_template' => 'results-:id-:datetime.csv',
```

According to T325177 the filename should be a bit more explicit, like so.

File: app/Http/Controllers/ImportsResultsController.php [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+class ImportsResultsController extends Controller
+{
+    public function showResultsCsv(int $import_id, ImportResults $results)
+    {
+        $import = ImportMeta::findOrFail($import_id);
+
+        $filename = strtr(config('imports.results.filename_template'), [
+            ':datetime' => now()->format('Y-m-dTH:i:s')
+        ]);
```
Comment (itamargiv) [inline app/Http/Controllers/ImportsResultsController.php:16]:
```suggestion
            ':id' => $import_id
            ':datetime' => now()->format('Y-m-dTH:i:s')
        ]);
```

In accordance with the file name template change suggested in the comment above.

File: app/Services/ImportResults.php [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+            return [
+                $mismatch['item_id'],
+                $mismatch['statement_guid'],
+                $mismatch['property_id'],
+                $mismatch['wikidata_value'],
+                $mismatch['meta_wikidata_value'],
+                $mismatch['external_value'],
+                $mismatch['external_url'],
+                $mismatch['review_status']
+            ];
```
Comment (itamargiv) [inline app/Services/ImportResults.php:33]: In a situation where we need to add a column, hard-coding the column keys here as well will require us to change two separate files.

However, if we map each key (as we get it from the config, into it's value here, we can avoid this redundancy)

File: app/Services/ImportResults.php [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+{
+    public function generateCSV(string $path, ImportMeta $import): void
+    {
+        $mismatches = Mismatch::get();
+
+        // Creating the output stream
+        $handle = fopen($path, 'w');
+
+        // Get the mismatches as a LazyCollection
+        $mismatches = Mismatch::whereBelongsTo($import)->cursor();
```
Comment (itamargiv) [inline app/Services/ImportResults.php:18]: Thanks for using a lazy collection, it's a good idea. Also, the comment makes it explicit so that we know which docs to refer to.

File: tests/Feature/ImportResultsTest.php [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+            return [
+                $mismatch['item_id'],
+                $mismatch['statement_guid'],
+                $mismatch['property_id'],
+                $mismatch['wikidata_value'],
+                $mismatch['meta_wikidata_value'],
+                $mismatch['external_value'],
+                $mismatch['external_url'],
+                $mismatch['review_status']
+            ];
```
Comment (itamargiv) [inline tests/Feature/ImportResultsTest.php:45]: Same as the prior comment, we can reduce redundancy here as well, so that we don't have to keep this updated in several files.

File: tests/Feature/WebStoreRouteTest.php [hunk 12]
```diff
@@ -64,4 +66,27 @@ public function test_store_import_csv_stats()
...
+    {
+        $response = $this->get(route('store.import-results', ['123123123']));
+        $response->assertStatus(404);
+    }
+
+    public function test_store_import_results_route()
+    {
+        $filename = strtr(config('imports.results.filename_template'), [
+            ':datetime' => now()->format('Y-m-dTH:i:s')
+        ]);
```
Comment (itamargiv) [inline tests/Feature/WebStoreRouteTest.php:80]:
```suggestion
            ':id' => $import->id
            ':datetime' => now()->format('Y-m-dTH:i:s')
        ]);
```

Same as the above comments regarding the change in the filename template.
ofc this will also require us to determine the ImportMeta prior to the filename assignment.

**Review by itamargiv — CHANGES_REQUESTED** [review]
Looks great! I've left a couple of comments, but besides that everything works great while running locally.

**Review by itamargiv — APPROVED** [review]
Great work! Thank you!
